### PR TITLE
Introduce expression API

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
Still WIP, working on tests.

### CODE REVIEW

**Comment by TolyaTalamanov** [comment]
@apertovs @l-bat @RodionovF @KSTconsult Have a look

File: src/layer_info.hpp [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+namespace deepworks {
+
+struct LayerInfo {
```
Comment (TolyaTalamanov) [inline src/layer_info.hpp:5]: CallInfo ?

**Comment by TolyaTalamanov** [comment]
> Still WIP, working on tests.

Done

**Comment by apertovs** [comment]
Looks good to me

**Review by apertovs — APPROVED** [review]

File: include/deepworks/shape.hpp [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#pragma once
+
+#include <vector>
+
+namespace deepworks {
+
+using Shape = std::vector<int>;
```
Comment (l-bat) [inline include/deepworks/shape.hpp:7]: Please rebase your branch to master

File: src/call.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <memory> // shared_ptr
+
+#include "layer_info.hpp"
+
+namespace deepworks {
+
+struct Placeholder;
```
Comment (l-bat) [inline src/call.hpp:9]:
```suggestion
classPlaceholder;
```

File: tests/unit/expression_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+#include "call.hpp"
+#include "layer_info.hpp"
+#include "placeholder_priv.hpp"
+#include "call_priv.hpp"
+
+namespace dw = deepworks;
+
+namespace {
+
+dw::Placeholder Foo(dw::Placeholder in,
```
Comment (l-bat) [inline tests/unit/expression_tests.cpp:19]: I'm not sure that's a good name

File: include/deepworks/placeholder.hpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+namespace deepworks {
+
+class Call;
+class Placeholder {
+public:
+    explicit Placeholder(const deepworks::Shape& shape);
+
+    const Shape& shape() const;
+
+    // NB: Public for test, but not available for user,
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply include/deepworks/placeholder.hpp:16]: Done

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -1,6 +1,20 @@
...
+set(HDR_FILES
+    "${PROJECT_SOURCE_DIR}/include/*.hpp"
+   )
 
 add_library(${PROJECT_NAME} SHARED ${SRC_FILES})
+# Set public API
+target_include_directories(${PROJECT_NAME}
+                           PUBLIC 
+                           ${HDR_FILES}
+                           # FIXME: Needed only for tests,
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply src/CMakeLists.txt:15]: Done

File: CMakeLists.txt [hunk 7]
```diff
@@ -10,12 +10,6 @@ project(deepworks)
 ########################################
 set(CMAKE_CXX_STANDARD 17)
 
-########################################
```
Comment (TolyaTalamanov) [inline CMakeLists.txt:13]: No global `include_directories`

File: include/deepworks/shape.hpp (hunk 2 above)
Comment (TolyaTalamanov) [reply include/deepworks/shape.hpp:7]: Move `Shape` to another file, to use it from placeholder

File: src/CMakeLists.txt [hunk 8]
```diff
@@ -1,7 +1,19 @@
-file(GLOB SRC_FILES ${PROJECT_SOURCE_DIR}/src/*.cpp)
+set(SRC_FILES
```
Comment (TolyaTalamanov) [inline src/CMakeLists.txt:1]: Let's keep it explicitly

File: tests/unit/expression_tests.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply tests/unit/expression_tests.cpp]: Removed tests

**Comment by TolyaTalamanov** [comment]
@l-bat

**Comment by apertovs** [comment]
Can we replace Prim to ClassNameImpl?

**Comment by TolyaTalamanov** [comment]
> Can we replace Priv with ClassNameImpl?

I believe `Class::Impl` would be enough

**Review by l-bat — APPROVED** [review]

**Pull request merged**